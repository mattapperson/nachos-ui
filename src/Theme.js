import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
