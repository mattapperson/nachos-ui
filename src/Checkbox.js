import React from 'react'
import PropTypes from 'prop-types'
import { View, TouchableOpacity } from 'react-native'
import Icon from 'react-native-vector-icons/dist/Ionicons'
import { withTheme } from './Theme'

const defaultTheme = {
  CHECKBOX_SIZE: 26,
  CHECKBOX_COLOR: '#fff',
  CHECKBOX_BORDER_WIDTH: 2,
  CHECKBOX_BORDER_COLOR: '#bdc1cc',
  CHECKBOX_CHECK_ICON_NAME: 'md-checkmark',
  CHECKBOX_CHECK_ICON_SIZE: 20,
  CHECKBOX_CHECK_ICON_COLOR: '#fff',
  CHECKBOX_CHECK_CIRCLE_COLOR: '#2f8cff',
  CHECKBOX_CHECKED_BG_COLOR: '#2f8cff',
  CHECKBOX_CHECKED_BORDER_COLOR: '#2f8cff',
  CHECKBOX_CIRCLE_RADIUS: 9999,
  CHECKBOX_ROUNDED_RADIUS: 5,
}

const Checkbox = (props) => {
  const {
    activeOpacity,
    disabled,
    kind,
    checked,
    onValueChange,
    style,
    disabledStyle,
    iconName,
    checkType,
    // NOTE: injected by a Switcher
    value,
    selected,
    onChange,
    theme,
  } = props


  // NOTE: function onChange is injected by the Switcher component
  const switcherProp = onChange &&
    {
      onPress: () => {},
      onPressOut: onChange.bind(null, value),
    }

  const isChecked = checked || selected

  let CheckComponent
  if (isChecked) {
    CheckComponent = <View style={[theme.check.circle]} />
    if (checkType === 'icon') {
      CheckComponent = (
        <Icon
          name={iconName || theme.settings.CHECKBOX_CHECK_ICON_NAME}
          size={theme.settings.CHECKBOX_CHECK_ICON_SIZE}
          color={theme.settings.CHECKBOX_CHECK_ICON_COLOR}
          style={theme.check.icon}
        />
      )
    }
  }

  const checkboxStyle = [
    theme.base,
    theme.kind[kind],
    isChecked && checkType === 'icon'
      ? theme.states.checkedIcon
      : theme.states.normal,
    style,
  ]

  return (
    <View style={disabled ? disabledStyle : {}}>
      <TouchableOpacity
        activeOpacity={activeOpacity}
        disabled={disabled}
        style={checkboxStyle}
        onPress={() => onValueChange(!checked)}
        {...switcherProp}
      >
        {CheckComponent}
      </TouchableOpacity>
    </View>
  )
}

Checkbox.defaultStyle = (theme = defaultTheme) => {
  return {
    settings: {
      CHECKBOX_CHECK_ICON_NAME: theme.CHECKBOX_CHECK_ICON_NAME,
      CHECKBOX_CHECK_ICON_SIZE: theme.CHECKBOX_CHECK_ICON_SIZE,
      CHECKBOX_CHECK_ICON_COLOR: theme.CHECKBOX_CHECK_ICON_COLOR,
    },
    base: {
      position: 'relative',
      width: theme.CHECKBOX_SIZE,
      height: theme.CHECKBOX_SIZE,
      borderWidth: theme.CHECKBOX_BORDER_WIDTH,
      justifyContent: 'center',
      alignItems: 'center',
    },
    kind: {
      circle: {
        // NOTE: we cannot use '50%' as we don't know the dimensions up front
        borderRadius: theme.CHECKBOX_CIRCLE_RADIUS,
      },
      rounded: {
        borderRadius: theme.CHECKBOX_ROUNDED_RADIUS,
      },
    },
    states: {
      normal: {
        borderColor: theme.CHECKBOX_BORDER_COLOR,
      },
      checkedIcon: {
        borderColor: theme.CHECKBOX_CHECKED_BORDER_COLOR,
        backgroundColor: theme.CHECKBOX_CHECKED_BG_COLOR,
      },
    },
    check: {
      icon: {
        backgroundColor: 'transparent',
        marginTop: 2,
      },
      circle: {
        width: 12,
        height: 12,
        borderRadius: 6,
        backgroundColor: theme.CHECKBOX_CHECK_CIRCLE_COLOR,
      },
    },
  }
}

Checkbox.propTypes = {
  activeOpacity: PropTypes.number,
  disabled: PropTypes.bool,
  kind: PropTypes.oneOf(['circle', 'rounded']),
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  disabledStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
  ]),
  checked: PropTypes.bool,
  checkType: PropTypes.oneOf(['icon', 'circle']),
  onValueChange: PropTypes.func,
  iconName: PropTypes.string,
  theme: PropTypes.object,
  // NOTE: injected by a Switcher
  value: PropTypes.string,
  onChange: PropTypes.func,
  selected: PropTypes.bool,
}

Checkbox.defaultProps = {
  activeOpacity: 0.8,
  disabled: false,
  kind: 'rounded',
  disabledStyle: { opacity: 0.3 },
  checked: false,
  onValueChange: () => {},
  checkType: 'icon',
}

export default withTheme('Checkbox', Checkbox)
