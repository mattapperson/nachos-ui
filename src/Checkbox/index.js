import React from "react";
import PropTypes from "prop-types";
import { View, TouchableOpacity } from "react-native";
import Icon from "react-native-vector-icons/dist/Ionicons";
import { withTheme } from "../Theme";

const Checkbox = props => {
  const {
    activeOpacity,
    disabled,
    checkComponent,
    checked,
    onValueChange,
    style,
    disabledStyle,
    iconName,
    iconColor,
    iconSize,
    // NOTE: injected by a Switcher
    value,
    selected,
    onChange,
    theme
  } = props;

  // NOTE: function onChange is injected by the Switcher component
  const switcherProp = onChange && {
    onPress: () => {},
    onPressOut: onChange.bind(null, value)
  };

  const isChecked = checked || selected || false;
  return (
    <View style={disabled ? disabledStyle : {}}>
      <TouchableOpacity
        activeOpacity={activeOpacity}
        disabled={disabled}
        style={[theme.base, isChecked ? theme.active : {}, style]}
        onPress={() => onValueChange(!checked)}
        {...switcherProp}
      >
        {isChecked &&
          !checkComponent && (
            <Icon
              name={iconName}
              size={iconSize}
              color={iconColor}
              style={theme.check}
            />
          )}
        {isChecked && checkComponent}
      </TouchableOpacity>
    </View>
  );
};

Checkbox.themeConfig = {
  props: {
    iconName: "md-checkmark",
    iconColor: "#fff",
    iconSize: 18
  },
  style: {
    base: {
      position: "relative",
      width: 22,
      height: 22,
      borderWidth: 2,
      justifyContent: "center",
      alignItems: "center",
      borderColor: "#ececee",
      borderRadius: 5
    },
    active: {
      backgroundColor: "#fff"
    },
    check: {
      backgroundColor: "transparent",
      color: "@primaryColor",
      marginTop: 2
    }
  }
};

Checkbox.propTypes = {
  activeOpacity: PropTypes.number,
  disabled: PropTypes.bool,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  disabledStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  checked: PropTypes.bool,
  onValueChange: PropTypes.func,
  iconName: PropTypes.string,
  theme: PropTypes.object.isRequired,
  // NOTE: injected by a Switcher
  value: PropTypes.string,
  onChange: PropTypes.func,
  selected: PropTypes.bool
};

Checkbox.defaultProps = {
  activeOpacity: 0.8,
  disabled: false,
  disabledStyle: { opacity: 0.3 },
  checked: false,
  onValueChange: () => {},
  checkType: "icon"
};

export default withTheme("Checkbox", Checkbox);
