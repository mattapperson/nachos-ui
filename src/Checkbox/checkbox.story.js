import { storiesOf } from "../storybook";
import { withInfo } from "../storybook";
import React from "react";
import Checkbox from "./index";
import { View } from "react-native";

const Example = () => {
  const checkboxStyle = { margin: 15 };

  return (
    <View style={{ marginVertical: 15 }}>
      <Checkbox style={checkboxStyle} checked={false} />
      <Checkbox style={checkboxStyle} kind="circle" checked={false} />
      <Checkbox style={checkboxStyle} disabled />
    </View>
  );
};

storiesOf("Checkbox", module).add("unchecked", withInfo()(Example));

const CheckedExample = () => {
  const checkboxStyle = { margin: 15 };

  return (
    <View style={{ marginVertical: 15 }}>
      <Checkbox style={checkboxStyle} checked={true} />
      <Checkbox style={checkboxStyle} kind="circle" checked={true} />
      <Checkbox style={checkboxStyle} checked disabled />
    </View>
  );
};

storiesOf("Checkbox", module).add("checked", withInfo()(CheckedExample));

export default Example;
