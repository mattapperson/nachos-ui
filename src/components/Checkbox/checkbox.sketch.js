import React from "react";
import Checkbox from "./index";

export const components = [
  {
    Component: () => <Checkbox checked={false} />,
    name: "Checkbox/unchecked/enabled"
  },
  {
    Component: () => <Checkbox checked={true} />,
    name: "Checkbox/checked/enabled"
  },
  {
    Component: () => <Checkbox checked={false} disabled />,
    name: "Checkbox/unchecked/disabled"
  },
  {
    Component: () => <Checkbox checked={true} disabled />,
    name: "Checkbox/checked/disabled"
  }
];
