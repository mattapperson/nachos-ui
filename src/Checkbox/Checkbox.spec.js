import "react-native";
import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";

import Checkbox from "../Checkbox";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Checkbox checked />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render disabled", () => {
  const wrapper = shallow(
    <Provider>
      <Checkbox disabled />
    </Provider>
  );
  expect(wrapper.find("TouchableOpacity").prop("disabled")).toBe(true);
});

it("should call onChange callback", () => {
  const callback = jest.fn();
  const wrapper = shallow(
    <Provider>
      <Checkbox onValueChange={callback} />
    </Provider>
  );
  wrapper.find("TouchableOpacity").simulate("press");
  expect(callback).toBeCalled();
});
