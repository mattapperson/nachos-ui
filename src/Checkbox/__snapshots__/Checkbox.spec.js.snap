// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1awozwy rn-backgroundColor-44z8sh rn-borderTopColor-w1rywy rn-borderRightColor-rieqnu rn-borderBottomColor-13kj4ez rn-borderLeftColor-1nlrok8 rn-borderTopLeftRadius-1as3g83 rn-borderTopRightRadius-dfv94e rn-borderBottomRightRadius-qpmf2f rn-borderBottomLeftRadius-re1h2s rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-1jxfwug rn-borderRightWidth-18p6if4 rn-borderBottomWidth-wgabs5 rn-borderLeftWidth-dwliz8 rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-height-18yzcnr rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-width-yc9v9c rn-zIndex-1lgpqti"
    disabled={false}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    tabIndex="0"
  >
    <div
      className="rn-backgroundColor-wib322 rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-color-q4xgo8 rn-display-1471scf rn-fontFamily-1t8snif rn-fontStyle-o11vmf rn-fontVariant-ebii48 rn-fontWeight-gul640 rn-lineHeight-t9a87b rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-marginTop-1bymd8e rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
      style={
        Object {
          "fontSize": "18px",
        }
      }
    >
      î—Š
    </div>
  </div>
</div>
`;

exports[`should render disabled 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
  style={
    Object {
      "opacity": 0.3,
    }
  }
>
  <div
    aria-disabled={true}
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1pxmb3b rn-flexBasis-7vfszb rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    disabled={true}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  />
</div>
`;
